libvast/vast/concept/parseable/core/rule.hpp

```diff
@@ -50,6 +95,55 @@ class rule_definition : public abstract_rule<Iterator, Attribute> {
 
 } // namespace detail
 
+/// A type-erased parser which can store any other parser. This type exhibits
+/// value semantics and can therefore not be used to construct recursive
+/// parsers.
+template <class Iterator>
+class erased_parser : public parser<erased_parser<Iterator>> {
```

Comment:
Why did you omit `typed_` in the name? Calling this class `erased_parser` left me wondering "erased... what?" while reading the code and I had to double check here that it's actually just type erasure.

Also, I miss a motivation to this class. It seems like this is functionally equivalent to `rule<>`, so what is the reason to have a new class in the first place? Only to avoid one indirection? If that's the case I would first want to see performance metrics that show a bottleneck with `rule<>`.

Reply:
> Also, I miss a motivation to this class. It seems like this is functionally equivalent to rule<>, so what is the reason to have a new class in the first place? Only to avoid one indirection? If that's the case I would first want to see performance metrics that show a bottleneck with rule<>.

This class is actually complementary to `rule<>`. `rule<>` parsers can be combined with other parsers through the `choice_parser` (`|`), but if you try to sequence it with `>>`, you actually end up with a parser that recurses infinitely. That is why I created the `erased_parser`, which cannot represent recursive grammars. It only provides type erasure, nothing else.

Reply:
> Why did you omit typed_ in the name? Calling this class erased_parser left me wondering "erased... what?" while reading the code and I had to double check here that it's actually just type erasure.

I'll rename it to `type_erased_parser`.